mps-sync-plugin/src/main/kotlin/org/modelix/mps/sync/plugin/gui/ModelSyncGuiFactory.kt

```diff
@@ -271,64 +274,82 @@ class ModelSyncGuiFactory : ToolWindowFactory, Disposable {
-                        populateBranchCB()
-                    }
+                val client = existingConnectionsModel.selectedItem as ModelClientV2
+                val repositories = client.listRepositories()
+
+                repoModel.removeAllElements()
+                repoModel.addAll(repositories)
+                if (repoModel.size > 0) {
+                    repoModel.selectedItem = repoModel.getElementAt(0)
+                    populateBranchCB()
                 }
             }
         }
 
-        private fun populateBranchCB() {
+        private suspend fun populateBranchCB() {
             if (existingConnectionsModel.size != 0 && repoModel.size != 0) {
-                CoroutineScope(Dispatchers.Default).launch {
-                    val branches =
-                        (existingConnectionsModel.selectedItem as ModelClientV2).listBranches(repoModel.selectedItem as RepositoryId)
-                    branchModel.removeAllElements()
-                    branchModel.addAll(branches)
-                    if (branchModel.size > 0) {
-                        branchModel.selectedItem = branchModel.getElementAt(0)
-                        populateModuleCB()
-                    }
+                val client = existingConnectionsModel.selectedItem as ModelClientV2
+                val repositoryId = repoModel.selectedItem as RepositoryId
+                val branches = client.listBranches(repositoryId)
+
+                branchModel.removeAllElements()
+                branchModel.addAll(branches)
+                if (branchModel.size > 0) {
+                    branchModel.selectedItem = branchModel.getElementAt(0)
+                    populateModuleCB()
                 }
             }
         }
 
-        private fun populateModuleCB() {
+        private suspend fun populateModuleCB() {
             if (existingConnectionsModel.size != 0 && repoModel.size != 0 && branchModel.size != 0) {
-                CoroutineScope(Dispatchers.Default).launch {
-                    val branch =
-                        (existingConnectionsModel.selectedItem as ModelClientV2).getReplicatedModel(branchModel.selectedItem as BranchReference)
-                            .start()
-                    branch.runRead {
-                        moduleModel.removeAllElements()
-                        val children = branch.getRootNode().allChildren.map {
-                            val name = it.getPropertyValue(BuiltinLanguages.jetbrains_mps_lang_core.INamedConcept.name)
-                                ?: it.toString()
-                            INodeWithName(it, name)
-                        }
-                        moduleModel.addAll(children.toList())
+                val client = existingConnectionsModel.selectedItem as ModelClientV2
+                val branchReference = branchModel.selectedItem as BranchReference
+                val branch = client.getReplicatedModel(branchReference).start()
```

Comment:
The `ReplicatedModel` here is not closed (and not used further).
Also fetching all data does not make sense here.

I would suggest to use a ModelQL query here.

Reply:
These replicated models are probably used at other places as well. Their life-cycle should be managed somewhere globally and then being reused.

Reply:
Definitely.  If I understood it correctly, the new plugin will only have one connection and one branch checked out per project. So there should be only one ReplicatedModel per project an its lifecycle should be bound to the project.

(I think I would not try to optimize reusing Clients/ReplicatedModels between opened projects to make it simple (and seperate failures).)

---

In this specific case, the user is still just exploring available branches and available modules in the UI.
For that, we definitely should not load the whole model and, for example, use ModelQL instead.

Reply:
fixed in [d4bb450](https://github.com/modelix/modelix.mps-plugins/pull/73/commits/d4bb45091b260acb9154445de5ca274ff4cb4c3e)